core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -158,17 +158,14 @@ class ZooKeeperClient(connectString: String,
           inReadLock(initializationLock) {
             send(request) { response =>
               responseQueue.add(response)
-              inFlightRequests.release()
-              countDownLatch.countDown()
             }
           }
-        } catch {
-          case e: Throwable =>
-            inFlightRequests.release()
-            throw e
+       } finally {
+          inFlightRequests.release()
+          countDownLatch.countDown()
         }
       }
-      countDownLatch.await()
+      countDownLatch.await(sessionTimeoutMs, TimeUnit.MILLISECONDS)
```

Comment:
@hachikuji @junrao 
What do you think of the timeout added here ?
With this addition, the above release of the coutdown latch is not strictly needed.